pkg/migrate/attributemigration.go

```diff
@@ -0,0 +1,336 @@
+	UpdateAPIV1ResourceInstance(url string, ri *v1.ResourceInstance) (*v1.ResourceInstance, error)
+	CreateSubResourceScoped(scopeKindPlural, scopeName, resKindPlural, name, group, version string, subs map[string]interface{}) error
+}
+
+type item struct {
+	ri     *v1.ResourceInstance
+	update bool
+}
+
+type migrateFunc func(ri *v1.ResourceInstance) error
+
+// AttributeMigration - used for migrating attributes to subresource
+type AttributeMigration struct {
+	client client
+	cfg    config.CentralConfig
+}
+
+// NewAttributeMigration creates a new AttributeMigration
+func NewAttributeMigration(client client, cfg config.CentralConfig) *AttributeMigration {
+	return &AttributeMigration{
+		client: client,
+		cfg:    cfg,
+	}
+}
+
+// Migrate - receives an APIService as a ResourceInstance, and checks if an attribute migration should be performed.
+// If a migration should occur, then the APIService, Instances, Revisions, and ConsumerInstances
+// that refer to the APIService will all have their attributes updated.
+func (m *AttributeMigration) Migrate(ri *v1.ResourceInstance) (*v1.ResourceInstance, error) {
+	if ri.Kind != mv1a.APIServiceGVK().Kind {
+		return ri, fmt.Errorf("expected resource instance kind to be api service")
+	}
+
+	// skip migration if x-agent-details is found for the service.
+	details := util.GetAgentDetails(ri)
+	if len(details) > 0 {
+		return ri, nil
+	}
+
+	funcs := []migrateFunc{
+		m.updateSvc,
+		m.updateRev,
+		m.updateInst,
+		m.updateCI,
+	}
+
+	errCh := make(chan error, len(funcs))
+	wg := &sync.WaitGroup{}
+
+	for _, f := range funcs {
+		wg.Add(1)
+
+		go func(fun migrateFunc) {
+			defer wg.Done()
+
+			err := fun(ri)
+			errCh <- err
+		}(f)
+	}
+
+	wg.Wait()
+	close(errCh)
+
+	for e := range errCh {
+		if e != nil {
+			return ri, e
+		}
+	}
+
+	return ri, nil
+}
+
+// updateSvc updates the attributes on service in place, then updates on api server.
+func (m *AttributeMigration) updateSvc(ri *v1.ResourceInstance) error {
+	url := fmt.Sprintf("%s/%s", m.cfg.GetServicesURL(), ri.Name)
+	r, err := m.getRI(url)
+	if err != nil {
+		return err
+	}
+	item := updateAttrs(r)
+	if !item.update {
+		return nil
+	}
+
+	// replace the address value so that the Migrate func can return the updated resource instance
+	*ri = *item.ri
+
+	return m.updateRI(url, item.ri)
+}
+
+// updateRev gets a list of revisions for the service and updates their attributes.
+func (m *AttributeMigration) updateRev(ri *v1.ResourceInstance) error {
+	url := m.cfg.GetRevisionsURL()
+	q := map[string]string{
+		"query": queryFunc(ri.Name),
+	}
+
+	return m.migrate(url, q)
+}
+
+// updateInst gets a list of instances for the service and updates their attributes.
+func (m *AttributeMigration) updateInst(ri *v1.ResourceInstance) error {
+	revURL := m.cfg.GetRevisionsURL()
+
+	q := map[string]string{
+		"query":  queryFunc(ri.Name),
+		"fields": "name",
+	}
+
+	revs, err := m.client.GetAPIV1ResourceInstancesWithPageSize(q, revURL, 100)
```

Comment:
I realized that the query that was being used was incorrect for api service instances. They only have a ref to a revision, so I updated this func to fetch all revisions for the service, and then update each instance for each revision.